tests/test_versions.py

```diff
@@ -0,0 +1,75 @@
+# This file is part of lsst_versions.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# Use of this source code is governed by a 3-clause BSD-style
+# license that can be found in the LICENSE file.
+
+import os
+import sys
+import tarfile
+import unittest
+
+try:
+    import git
+except ImportError:
+    git = None
+
+from lsst_versions import find_dev_lsst_version
+
+TESTDIR = os.path.abspath(os.path.dirname(__file__))
+GITDIR = os.path.join(TESTDIR, "repo")
+TARFILE = os.path.join(TESTDIR, "test-repo.tgz")
+
+
+def setup_module(module):
+    """Ensure that the test git repository is present.
+
+    This repository is stored as a tar file and must be unpacked
+    before the tests run.
+    """
+    if not os.path.exists(GITDIR):
+        with tarfile.open(TARFILE, "r:gz") as tar:
+            tar.extractall(path=TESTDIR)
+
+
+@unittest.skipIf(git is None, "GitPython package is not installed.")
+class VersionsTestCase(unittest.TestCase):
+    """Test Git version finding."""
+
+    def setUp(self):
+        try:
+            git.Repo(GITDIR)
+        except Exception:
+            raise unittest.SkipTest("Git repository for this package is not accessible.")
+
+    def test_versions(self):
+        """Determine versions of a test repository."""
+        versions = (
+            ("86427e5", "1.0.0a00000000"),  # No parents
+            ("86b5d01", "1.0.0a00000001"),
+            ("595e858", "1.0"),
+            ("ea28756", "2.0.0a20220400"),
+            ("af0c308", "2.0.0a20220100"),
+            ("w.2022.1", "2.0.0a20220100"),
+            ("da7a09d", "2.0.0a20220401"),
+            ("v2.1.0", "2.1.0"),
+            ("w.2022.05", "2.0.0a20220700"),
```

Comment:
Huh?  Not 2.0.0a20220500?

Reply:
Well, this is the repo:

```
* 8d31d74 - (tag: w.2022.7, tag: w.2022.06, tag: w.2022.05)
```

and it converts the tag to a commit and then the commit gets converted back to a tag with priority given to the most recent tag. You make a compelling case that in order to make the version invariant as newer tags are added it should always pick the oldest weekly... (it will pick the most recent release though -- so for 23.0.0 and 23.0.1 it will pick 23.0.1).

Or are you saying that if the user specifies a weekly tag explicitly then we should short circuit all the cleverness and use that weekly directly?

Reply:
My initial comment came from not understanding how 5 became 7.  After I figured it out, I wasn't bothered.

I believe pkgautoversion tried to do the oldest version for invariance, but that can also be confusing.  Your call.

Using the user's string (assuming that weekly exists) is interesting, but I think it's a very rare use case and so not really worth worrying about.